### PR TITLE
Feature/issues/408 2

### PR DESCRIPTION
**Pull request by Mizzick**
#408 Collapser shadowDOM support

### CODE REVIEW

File: Extension/lib/content-script/preload.js [hunk 1]
```diff
@@ -57,9 +57,12 @@
         // We use shadow DOM when it's available to minimize our impact on web page DOM tree.
         // According to ABP issue #452, creating a shadow root breaks running CSS transitions.
         // Because of this, we create shadow root right after content script is initialized.
-        if ("createShadowRoot" in document.documentElement && shadowDomExceptions.indexOf(document.domain) == -1) {
-            shadowRoot = document.documentElement.createShadowRoot();
-            shadowRoot.appendChild(document.createElement("shadow"));
+        var shadowRoot = document.documentElement.shadowRoot;
```
Comment (ameshkov) [inline Extension/lib/content-script/preload.js]: This is redundant

File: Extension/lib/utils/element-collapser.js [hunk 2]
```diff
@@ -152,25 +194,54 @@ var ElementCollapser = (function() {
...
     };
 
     /**
+     * Creates selector for specified tagName and src attribute
+     */
+    var createSelectorForSrcAttr = function(srcAttrValue, tagName) {
+        return tagName + '[src="'+ CSS.escape(srcAttrValue) + '"]';
+    };
+
+    var getStyleElement = function(shadowRoot) {
```
Comment (ameshkov) [inline Extension/lib/utils/element-collapser.js]: Comment?

File: Extension/lib/utils/element-collapser.js [hunk 3]
```diff
@@ -52,37 +54,83 @@ var ElementCollapser = (function() {
...
+    var unhideBySelector = function(selectorText, shadowRoot) {
+        var styleElement = getStyleElement(shadowRoot);
         if (!styleElement || !styleElement.sheet) {
             return;
         }
         var iLength = styleElement.sheet.cssRules.length;
         while (iLength--) {
             var cssRule = styleElement.sheet.cssRules[iLength];
-            if (cssRule.selectorText == selectorText) {
+            if (cssRule.selectorText == (shadowRoot ? '::content ' + selectorText : selectorText)) {
```
Comment (ameshkov) [inline Extension/lib/utils/element-collapser.js]: Can be moved to a separate helper function `prepareSelectorText(selectorText, shadowRoot)`

File: Extension/lib/utils/element-collapser.js [hunk 4]
```diff
@@ -152,25 +194,54 @@ var ElementCollapser = (function() {
...
      * @param element Element to collapse
+     * @param shadowRoot optional
      */
-    var collapseElement = function(element) {
+    var collapseElement = function(element, shadowRoot) {
 
         var tagName = element.tagName.toLowerCase();
+        var source = element.getAttribute('src');
+        if (source) {
+            //To not to keep track of changing src for elements, we gonna collapse it if special selector
```
Comment (ameshkov) [inline Extension/lib/utils/element-collapser.js]: we are going to collapse it with a CSS rule

**Comment by Mizzick** [comment]
Fixed

File: Extension/lib/content-script/preload.js [hunk 5]
```diff
@@ -57,7 +57,8 @@
         // We use shadow DOM when it's available to minimize our impact on web page DOM tree.
         // According to ABP issue #452, creating a shadow root breaks running CSS transitions.
         // Because of this, we create shadow root right after content script is initialized.
-        var shadowRoot = document.documentElement.shadowRoot;
+        // First check if it's available already
```
Comment (ameshkov) [inline Extension/lib/content-script/preload.js]: You should alert reader, that browser will print warning if you try to create additional root.

File: Extension/lib/utils/element-collapser.js [hunk 6]
```diff
@@ -224,6 +226,17 @@ var ElementCollapser = (function() {
     };
 
     /**
+     * Prepares selector or rule text
+     *
+     * @param selector
+     * @param useShadowDom
+     * @returns {*}
+     */
+    var prepareSelector = function (selector, useShadowDom) {
```
Comment (ameshkov) [inline Extension/lib/utils/element-collapser.js]: You should either pass `!!shadowRoot` as a parameter, or rename parameter to `shadowRoot` and use `return !!shadowRoot ? "::content " + selector : selector`

**Comment by Mizzick** [comment]
Ok

**Pull request merged**